subprojects/core/src/main/java/org/gradle/caching/configuration/internal/BuildCacheServiceRegistration.java

```diff
@@ -14,26 +14,12 @@
  * limitations under the License.
  */
 
-package org.gradle.caching;
+package org.gradle.caching.configuration.internal;
 
-import org.gradle.api.Incubating;
+import org.gradle.caching.BuildCacheService;
+import org.gradle.caching.configuration.BuildCache;
 
-/**
- * Factory interface to be provided by build cache service implementations.
- *
- * @param <T> the type of build cache configuration this factory can handle.
- *
- * @since 3.5
- */
-@Incubating
-public interface BuildCacheServiceFactory<T extends org.gradle.caching.configuration.BuildCache> {
-    /**
-     * Returns the type of the supported build cache configuration.
-     */
+public interface BuildCacheServiceRegistration<T extends BuildCache, S extends BuildCacheService> {
     Class<T> getConfigurationType();
```

Comment:
If we are going in this direction, it'd be nice to have two methods here, one for a public and one for an internal type, the internal type defaulting to the public type. It's a pity we can't use Java 8's default methods here. :/